aten/src/ATen/core/LazyTensor.cpp

```diff
@@ -0,0 +1,142 @@
+    return v;
+  }
+  auto fs = torch::jit::parseSchema(lt->schema_);
+  auto s = Symbol::fromQualString(fs.name());
+  auto* node = g->create(s);
+  auto* output = node->output();
+
+  auto i = 0;
+  for (const auto& inp_ : lt->inps_) {
+    torch::jit::Value* v;
+    if (inp_.isTensor()) {
+      auto t = inp_.toTensor();
+      if (c10::tensorTypeIdToBackend(t.type_id()) == Backend::Lazy) {
+        auto lt_2 = static_cast<LazyTensorImpl*>(t.unsafeGetTensorImpl());
+        v = getValue(lt_2, g, value_map);
+      } else {
+        v = g->addInput();
+        value_map[v] = inp_;
+      }
+    } else {
+      if (inp_.isNone()) {
+        auto& arg = fs.arguments()[i];
+        v = g->insertNode(g->createNone())->output();
+      } else {
+        v = insertConstant(*g, inp_);
+      }
+      value_map[v] = inp_;
+    }
+    node->addInput(v);
+    i++;
+  }
+  g->insertNode(node);
+  return output;
+}
+
+Tensor LazyTensorImpl::to_eager() const {
+  if (!schema_) {
+    return inps_[0].toTensor();
+  }
+
+  auto g = std::make_shared<torch::jit::Graph>();
+  std::map<torch::jit::Value*, c10::IValue> value_map;
+
+  auto* output = getValue(this, g, value_map);
+  torch::jit::Stack stack;
+  bool requires_grad = false;
+  for (auto& inp : g->inputs()) {
+    stack.push_back(value_map[inp]);
+    if (!value_map[inp].isTensor()) {
+      continue;
+    }
+    const auto& var = static_cast<const torch::autograd::Variable&>(
+        value_map[inp].toTensor());
+    if (var.defined() && var.requires_grad()) {
+      requires_grad = true;
+    }
+  }
+  g->registerOutput(output);
+
+  if (!requires_grad) {
+    torch::jit::InterpreterState(torch::jit::Code(g)).run(stack);
+    auto output_t = stack[0].toTensor();
+    if (output_t.is_variable()) {
+      return output_t;
+    }
+    auto output_var = torch::autograd::make_variable(output_t);
+    output_t = output_var.tensor_data();
+    TORCH_INTERNAL_ASSERT(output_t.is_variable());
+    return output_t;
+  }
+
+  auto grad = torch::jit::differentiate(g);
+  auto d = torch::jit::DifferentiableGraphOp(grad);
+  d(stack);
+
+  auto output_t = stack.at(0).toTensor();
+  if (output_t.is_variable()) {
+    return output_t;
+  }
+  auto output_var = torch::autograd::make_variable(output_t);
+  return output_var.tensor_data();
+}
+
+} // namespace at
+
+namespace at {
+
+CAFFE2_API at::Tensor single_output(
+    const char* schema,
+    std::vector<c10::IValue> const& inps) {
+  auto self = inps[0].toTensor();
+  auto tensor = detail::make_tensor<LazyTensorImpl>(
+      schema, inps, self.dtype(), self.device());
+  return torch::autograd::make_variable(tensor);
+}
+
+namespace native {
+
+CAFFE2_API at::Tensor eager_to_lazy(at::Tensor const& self) {
+  auto tensor = detail::make_tensor<LazyTensorImpl>(self);
+  return tensor;
+}
+
+CAFFE2_API at::Tensor lazy_to_eager(at::Tensor const& self) {
+  auto lt = static_cast<LazyTensorImpl*>(self.unsafeGetTensorImpl());
```

Comment:
I know in this particular case it is safe, but unchecked static casts always make me nervous (probably because I'm afraid someone will copy paste this code to a place where there the precondition that a variable is a tensor is NOT satisfied.)